File: src/rosHWpub.cpp
#include "ros/ros.h"
#include "std_msgs/String.h"
#include <sstream>

int main(int argc, char **argv)
{
  ros::init(argc, argv, "cpp_talker");
  ros::NodeHandle n;

  // First publisher
  ros::Publisher chatter_pub = n.advertise<std_msgs::String>("chatter", 1000);

  // Second publisher for a new topic
  ros::Publisher another_chatter_pub = n.advertise<std_msgs::String>("another_chatter", 1000);

  ros::Rate loop_rate(10);

  while (ros::ok())
  {
    // First message
    std_msgs::String msg1;
    std::stringstream ss1;
    ss1 << "cppPUB1 " << ros::Time::now();
    msg1.data = ss1.str();
    ROS_INFO("%s", msg1.data.c_str());
    chatter_pub.publish(msg1);

    // Second message
    std_msgs::String msg2;
    std::stringstream ss2;
    ss2 << "cppPUB2 " << ros::Time::now();
    msg2.data = ss2.str();
    ROS_INFO("%s", msg2.data.c_str());
    another_chatter_pub.publish(msg2);

    ros::spinOnce();
    loop_rate.sleep();
  }

  return 0;
}


// #include "ros/ros.h"
// #include "std_msgs/String.h"
// #include <sstream>

// int main(int argc, char **argv)
// {
//   ros::init(argc, argv, "cpp_talker");
//   ros::NodeHandle n;
//   ros::Publisher chatter_pub = n.advertise<std_msgs::String>("chatter", 1000);
//   ros::Rate loop_rate(10);

//   while (ros::ok())
//   {
//     std_msgs::String msg;
//     std::stringstream ss;
//     ss << "HW1 " << ros::Time::now();
//     msg.data = ss.str();

//     ROS_INFO("%s", msg.data.c_str());
//     chatter_pub.publish(msg);

//     ros::spinOnce();
//     loop_rate.sleep();
//   }

//   return 0;
// }



// #include "ros/ros.h"
// #include "std_msgs/String.h"
// #include <sstream>

// int main(int argc, char **argv)
// {
//   ros::init(argc, argv, "cpp_talker");
//   ros::NodeHandle n;
//   ros::Publisher chatter_pub = n.advertise<std_msgs::String>("topic", 1000);
//   ros::Rate loop_rate(10);

//   while (ros::ok())
//   {
//     std_msgs::String msg;
//     std::stringstream ss;
//     ss << "hello world " << ros::Time::now();
//     msg.data = ss.str();

//     ROS_INFO("%s", msg.data.c_str());
//     chatter_pub.publish(msg);
//     ros::spinOnce();
//     loop_rate.sleep();
//   }

//   return 0;
// }

File: script/rosHWpub.py
# 發布兩個話題
#!/usr/bin/env python3
import rospy
from std_msgs.msg import String

def talker():
    pub1 = rospy.Publisher('topic1', String, queue_size=10)
    pub2 = rospy.Publisher('topic2', String, queue_size=10)
    rospy.init_node('py_talker', anonymous=True) ###創建一個名為 py_talker 的 ROS 節點
    rate = rospy.Rate(10) # 10hz

    while not rospy.is_shutdown():
        msg1 = "Hello on topic 1 %s" % rospy.get_time()
        msg2 = "Hello on topic 2 %s" % rospy.get_time()

        rospy.loginfo(msg1)
        pub1.publish(msg1)

        rospy.loginfo(msg2)
        pub2.publish(msg2)

        rate.sleep()

if __name__ == '__main__':
    try:
        talker()
    except rospy.ROSInterruptException:
        pass




## 發布一個話題
## #!/usr/bin/env python3
## import rospy
## from std_msgs.msg import String

## def talker():
##     pub = rospy.Publisher('HW1', String, queue_size=10) #節點以10hz的頻率發布到chatter話題
##     rospy.init_node('py_talker', anonymous=True) #節點名稱為py_talker
##     rate = rospy.Rate(10) # 10hz
##     while not rospy.is_shutdown():
##         str = "S %s" % rospy.get_time()
##         rospy.loginfo(str)
##        pub.publish(str)
##         rate.sleep()

# if __name__ == '__main__':
#     try:
#         talker()
#     except rospy.ROSInterruptException:
#         pass
